resources/splunk_service.rb

```diff
@@ -0,0 +1,158 @@
+# Cookbook Name:: cerner_splunk_ingredient
+# Resource:: splunk_service
+#
+# Resource for managing Splunk as a system service
+class SplunkService < ChefCompat::Resource
+  include CernerSplunk::PlatformHelpers, CernerSplunk::ResourceHelpers,
+          CernerSplunk::ServiceHelpers, CernerSplunk::RestartHelpers
+
+  resource_name :splunk_service
+
+  property :name, String, name_property: true, identity: true
+  property :package, [:splunk, :universal_forwarder], required: true
+  property :options, String, default: ''
+  property :user, String, desired_state: false
+  property :ulimit, Integer, desired_state: false
+
+  default_action :nothing
+
+  def after_created
+    package_from_name unless property_is_set? :package
+  end
+
+  def process_name
+    @process_name ||= process_names[package]
+  end
+
+  action_class do
+    def execute_cmd(cmd, run_as = nil)
+      cmd_str = node['os'] == 'windows' ? "splunk.exe #{cmd}" : "./splunk #{cmd}"
+      cmd_str += " #{options}" unless options.empty?
+      cmd_str = "sh -c 'ulimit -n #{ulimit} && #{cmd_str}'" if ulimit
+
+      # Execute this resource with run_action, preventing race condition with write_initd_ulimit
+      execute cmd_str do
+        action :nothing
+        cwd splunk_bin_path.to_s
+        user run_as if run_as
+        live_stream true if defined? live_stream
+      end.run_action :run
+    end
+
+    def ensure_license_agreement(run_as)
+      execute_cmd('version --accept-license', run_as) if ftr_pathname(install_dir).exist?
+    end
+  end
+
+  ### Inherited Actions
+
+  load_current_value do |desired|
+    package desired.package
+
+    raise 'Attempted to reference service for Splunk installation that does not exist' unless load_installation_state
+    check_restart unless defined?(performed_actions) && !performed_actions.empty?
+
+    user current_owner
+
+    if node['os'] == 'windows'
+      desired.ulimit = nil
+      reset_property :ulimit
+    else
+      limits = service_limits
+      if service_running && limits
+        proc_limit = limits[/^Max open files \s+ \w+ \s+ (\w+) \s+ files\s*$/m, 1]
+        ulimit proc_limit == 'unlimited' ? -1 : proc_limit.to_i
+      else
+        init_file = Pathname.new('/etc/init.d/splunk')
+        limit = if init_file.exist?
+                  init_file.read[/ulimit -n (\d+)/, 1].to_i
+                else
+                  shell_out('sh -c "ulimit -n"', user: user).stdout.to_i
+                end
+        ulimit limit
+      end
+    end
+  end
+
+  action :start do
+    ensure_license_agreement user
+    execute_cmd('start', user)
```

Comment:
Ahh, I'm not as crazy as I thought. So here's an interesting case. Splunk has to be started through the init script or else the service can't be stopped properly. Seems to be dependent on this `/var/lock/subsys/splunk` file existing.

The implementation of checking for this file doesn't reside in the init script itself, so that's kind of a bummer. I was hoping to add another condition somewhere to check for the pid file. I can probably remedy this by touching the lock file if I detect the service is running when enable is called.